src/main/java/minijava/ir/IREmitter.java

```diff
@@ -0,0 +1,769 @@
+   */
+  private final IdentityHashMap<LocalVariable, Integer> localVarIndexes = new IdentityHashMap<>();
+  /** The Basic Block graph of the current function. */
+  private Graph graph;
+  /**
+   * Construction is a firm Node factory that makes sure that we don't duplicate expressions, thus
+   * making common sub expressions irrepresentable.
+   */
+  private Construction construction;
+  /**
+   * Stores the node's value in the current lvar (if there is any). This is crucial for assignment
+   * to work. E.g. in the expression x = 5, we analyze the variable expression x and get back its
+   * value, which is irrelevant for assignment, since we need the *address of* x. This is where we
+   * need storeInCurrentLval, which would store the expression node of the RHS (evaluating to 5) in
+   * the address of x.
+   *
+   * <p>Now, in an ideal world, this variable would be 'Node currentLval', but firm doesn't offer a
+   * function for getting the address of a local variable as a Node. So in order to not duplicate a
+   * lot of work (e.g. computing array offsets, etc.) in a mechanism without this variable, we
+   * abstract actual assignment out into a function.
+   */
+  @Nullable private Function<Node, Node> storeInCurrentLval;
+
+  public static void main(String[] main_args) {}
+
+  @Override
+  public Void visitProgram(minijava.ast.Program that) {
+    classTypes.clear();
+    methods.clear();
+    fields.clear();
+    for (Class decl : that.declarations) {
+      ClassType classType = new ClassType(decl.name());
+      classTypes.put(decl, classType);
+      for (Field f : decl.fields) {
+        fields.put(f, addFieldDecl(f));
+      }
+      if (decl.fields.size() == 0) {
+        // We have to prevent class types of length 0, so we insert an unreachable field.
+        String unusableFieldName = "0padding"; // unusable because of digit prefix
+        Entity fieldEnt = new Entity(classType, unusableFieldName, BOOLEAN_TYPE);
+        fieldEnt.setLdIdent(NameMangler.mangleInstanceFieldName(decl.name(), unusableFieldName));
+      }
+      for (Method m : decl.methods) {
+        methods.put(m, addMethodDecl(m));
+      }
+      //System.out.println("# " + classType);
+      //System.out.println(classType.getSize());
+      classType.layoutFields();
+      classType.finishLayout();
+      //System.out.println(classType.getSize());
+    }
+    for (Class klass : that.declarations) {
+      klass.methods.forEach(this::emitBody);
+    }
+    return null;
+  }
+
+  private Entity addFieldDecl(Field f) {
+    Type type = getStorageType(f.type);
+    ClassType definingClass = classTypes.get(f.definingClass.def);
+    Entity fieldEnt = new Entity(definingClass, f.name(), type);
+    fieldEnt.setLdIdent(NameMangler.mangleInstanceFieldName(definingClass.getName(), f.name()));
+    return fieldEnt;
+  }
+
+  /**
+   * Returns the primitive type in which to store type @type@. Consider a @boolean b;@ declaration,
```

Comment:
Guilty, I thought Javadoc was less clumsy. Should read up on that, but I think I just use backticks rather than worsening my RSI by typing that kind of abomination.